Blockbench project file parser rework (5.0 support)

penguinencounter opened the pull request:
Fixes #445.
Adds 5.0 project file support.

## 4.12
- [x] Load model
- [x] Load textures
- [x] Load animations
- [x] Load collections
- [x] BUG: UVs don't work
- [x] BUG: emissives don't work
- [x] Validate output equivalence to previous parser
- [x] Pass parity check w/ avatar-downloads
- [x] Convert null objects & locators to groups
- [x] Todo: handle `null` texture source

## 5.0
- [x] Load model
- [x] Load textures
- [x] Load animations
  - [x] Detect Lua corruption from 4.12 to 5.0 upgrade
  - [x] Negate numbers
  - [x] Negate expressions (wrap with `-( expr )`)
  - [x] Don't negate statements (note: see https://www.lua.org/manual/5.2/manual.html#9)
  - [x] Bonus: Make keyframe errors not suck
        new format is `<animname> keyframe (part '<partname>', time 1.23s, ROTATION X)`
- [x] Load collections
- [x] Pass parity check on converted avatars
  - [x] Validate NBT same-as 4.12 (or equivalent)
- [x] Pass check on new avatars (created in 5.0)
- [x] Convert null objects & locators to groups
- [x] BUG: Rotations are messed up (see Total)
- [x] Check smooth shading works
- [x] BUG: visib cascade workn't
- [x] BUG: https://discord.com/channels/1129805506354085959/1129815148190761081/1431290276957130872

## housekeeping
- [x] ~~Delete~~ Deprecate old parser
- [x] Remove fuzzer
- [x] Remove old parser for real this time
- [x] Determine what to do with armatures / bones / billboards
      Determination made: ignore it lol

penguinencounter committed:
Reworked parser first draft

penguinencounter committed:
Blockbench <=4.9 compatibility in new parser

penguinencounter committed:
minor: more clear log messages

penguinencounter committed:
Fixed a load of bugs.

- Textures were in hash order in NBT (instead of specification order), resulting in the wrong textures all over the place
- Bad faces on meshes crashed the game
- Visibility is now inherited better (parity)

penguinencounter committed:
Fix emissive textures not working

penguinencounter committed:
Oh no.

Animations are completely messed up.

- we need to negate some of the values
- blockbench might have messed up some users' Lua if they do the upgrade (but new code shouldn't have this issue)

penguinencounter committed:
Potentially mitigate the issue.

penguinencounter committed:
try at least a little with the Lua

penguinencounter committed:
Fix load fail with blank animators

penguinencounter committed:
Fix a bunch of issues that the fuzzer found

penguinencounter committed:
lint

penguinencounter committed:
Add fuzzer gradle task

penguinencounter committed:
Change how properties are mutated

penguinencounter committed:
Parse collections

penguinencounter committed:
Improve animation errors

- Tangentially related to this PR
- Definitely a step in the right direction
- Also, better detection for statement/expression
- And a warning for when it gets messed up

penguinencounter committed:
Don't try to write elements of unknown type into nbt

penguinencounter committed:
lint

penguinencounter committed:
Fuzzer updates (strings, optimization)

penguinencounter committed:
Clean up

- rm fuzzer & tasks
- rm testing system properties
- rm old parser

penguinencounter committed:
Allow point elements to have parent types

Also fixes them not being positioned properly, probably

penguinencounter committed:
Fix visibility propagation for v5 models

Can't believe I missed that lol

penguinencounter committed:
Gracefully handle `null` `source`

UnlikePaladin reviewed (approved):
Looks pretty solid overall, no complaints from me!

penguinencounter committed:
Switch from JOML to in-house types

Should make backporting easier.

penguinencounter committed:
Fix instruction keyframe crash

(omnibus PR at this point LUL)

penguinencounter committed:
packing: fixed detection of default bezier timings

penguinencounter committed:
elide `chld` if no children

penguinencounter committed:
Don't fail on `null` in vectors

This usually indicates corruption (NaN/Infinity) but Blockbench reads it as-if 0 so we'll do that too

penguinencounter committed:
add `loadOptions` field to meta

currently only contains a switch to include UUIDs with ModelParts.
rationale:
- some people probably want it (it got PRed for a reason, so at least nathan wants it presumably)
- most people don't
- it just takes up 16 bytes per part and is basically incompressible
result:
- off by default, opt in

penguinencounter committed:
loading avatars is deterministic again

we may or may not have thought "oh, why are they using a linked set?"

we now know why they used a linked hash set.

penguinencounter committed:
Format version field

penguinencounter committed:
better syntax error messages!

they're quite tall though

penguinencounter committed:
Remove newline at end of message

penguinencounter committed:
Report a dependency error if v5addon is installed

we're trying to save people from the mixin error that will inevitably happen if they leave the addon installed
featuring the worst version range the world has ever seen

penguinencounter commented:
lexforge lets you get away with some complete BS in dependency version strings
```toml
versionRange = "[NONE - §cincompatible§r\\nThis addon's features are included in this Figura version!]"
```

<img width="978" height="160" alt="image" src="https://github.com/user-attachments/assets/3b206c98-5a2e-4e2b-822b-50ee1b1f659d" />

penguinencounter committed:
Merge branch '1.20' into bbv5
